File: java/pe015.java
//file: pe015.java

import java.util.*;
import java.math.BigInteger;

public class pe015 {
  public static void main( String[] args ) {
    final int max = 20;
//  need to compute 2n!/((n!)*(n!))
    ArrayList<Integer> numerator = new ArrayList<Integer>();
    ArrayList<Integer> denominator = new ArrayList<Integer>();
    for (int i = 0; i < max; i++) {
      denominator.add(i+1);
      numerator.add(i+max+1);
    }

/*
  walk through lists and cast away multiples to reduce size of numbers from
  factorials - this is incomplete factoring but sufficient for this problem
  we could factor both completely if we wanted...
*/

    for (int i=0; i < max; i++) {
      for (int j=0; j < max; j++) {
        if ((denominator.get(i) != 1) && ((numerator.get(j) % denominator.get(i)) == 0)) {
          numerator.set(j, (numerator.get(j) / denominator.get(i)));
          denominator.set(i, 1);
        }
      }
    }

    BigInteger num = BigInteger.ONE;
    BigInteger denom = BigInteger.ONE;
    for (int i = 0; i < max; i++) {
      num = num.multiply(BigInteger.valueOf(numerator.get(i)));
      denom = denom.multiply(BigInteger.valueOf(denominator.get(i)));
    }
  
    System.out.println((num.divide(denom)).longValue());
  }
}

File: typescript/pe015.ts
function pe015(max15:number): number {
  // need to compute 2n!/((n!)*(n!))

  let numerator = new Array(max15)
  let denominator = new Array(max15)

  for (let i = 0; i < max15; i++)
  {
    denominator[i] = i+1
    numerator[i] = i+max15+1
  }

  // walk through lists and cast away multiples to reduce size of numbers from factorials

  for (let i=0; i < max15; i++)
  {
    for (let j=0; j < max15; j++)
    {
      if ((denominator[i] != 1) && (!(numerator[j] % denominator[i])))
      {
        numerator[j] = numerator[j] / denominator[i]
        denominator[i] = 1
      }
    }
  }

  let num = 1
  let denom = 1
  for (let i=0; i < max15; i++)
  {
    num *= numerator[i]
    denom *= denominator[i]
  }

  return num/denom
}

const max15 = 20
console.log(pe015(max15))
